File: nodejs/express5-postgres/src/telemetry.ts
/**
 * OpenTelemetry instrumentation setup for Express 5 application.
 *
 * CRITICAL: This file MUST be imported before any other modules
 * to ensure auto-instrumentation captures all dependencies.
 */

import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { OTLPLogExporter } from '@opentelemetry/exporter-logs-otlp-http';
import { Resource } from '@opentelemetry/resources';
import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { BatchLogRecordProcessor } from '@opentelemetry/sdk-logs';
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
} from '@opentelemetry/semantic-conventions';

const otlpEndpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'http://localhost:4318';
const serviceName = process.env.OTEL_SERVICE_NAME || 'express5-postgres-app';
const serviceVersion = process.env.OTEL_SERVICE_VERSION || '1.0.0';

const resource = new Resource({
  [ATTR_SERVICE_NAME]: serviceName,
  [ATTR_SERVICE_VERSION]: serviceVersion,
});

const sdk = new NodeSDK({
  resource,
  traceExporter: new OTLPTraceExporter({
    url: `${otlpEndpoint}/v1/traces`,
  }),
  metricReader: new PeriodicExportingMetricReader({
    exporter: new OTLPMetricExporter({
      url: `${otlpEndpoint}/v1/metrics`,
    }),
    exportIntervalMillis: 60000,
  }),
  logRecordProcessor: new BatchLogRecordProcessor(
    new OTLPLogExporter({
      url: `${otlpEndpoint}/v1/logs`,
    })
  ),
  instrumentations: [
    getNodeAutoInstrumentations({
      '@opentelemetry/instrumentation-http': {
        ignoreIncomingRequestHook: (req) => {
          const url = req.url || '';
          return url === '/api/health' || url === '/health';
        },
      },
      '@opentelemetry/instrumentation-fs': {
        enabled: false,
      },
    }),
  ],
});

sdk.start();

process.on('SIGTERM', () => {
  sdk
    .shutdown()
    .then(() => console.log('OpenTelemetry SDK shut down'))
    .catch((err) => console.error('Error shutting down OpenTelemetry SDK', err))
    .finally(() => process.exit(0));
});

export { sdk };

File: python/flask-postgres/app/telemetry.py
"""OpenTelemetry instrumentation setup for Flask application.

Configures traces, metrics, and logs with OTLP exporters.
"""

import logging
import os

from opentelemetry import _logs, metrics, trace
from opentelemetry.exporter.otlp.proto.http._log_exporter import OTLPLogExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.celery import CeleryInstrumentor
from opentelemetry.instrumentation.flask import FlaskInstrumentor
from opentelemetry.instrumentation.logging import LoggingInstrumentor
from opentelemetry.instrumentation.redis import RedisInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource, get_aggregated_resources
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


_otel_log_handler: LoggingHandler | None = None
_initialized: bool = False


def setup_telemetry() -> None:
    """Initialize OpenTelemetry with traces, metrics, and logs.

    This function should be called once at application startup,
    BEFORE creating the Flask app.
    """
    global _otel_log_handler, _initialized

    if _initialized:
        return

    otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318")
    service_name = os.getenv("OTEL_SERVICE_NAME", "flask-postgres-app")
    service_version = os.getenv("OTEL_SERVICE_VERSION", "1.0.0")

    # Create resource - get_aggregated_resources automatically picks up OTEL_RESOURCE_ATTRIBUTES
    resource = get_aggregated_resources(
        detectors=[],
        initial_resource=Resource.create(
            {
                "service.name": service_name,
                "service.version": service_version,
            }
        ),
    )

    # ==========================================================================
    # 1. Traces
    # ==========================================================================
    trace_exporter = OTLPSpanExporter(endpoint=f"{otlp_endpoint}/v1/traces")
    trace_provider = TracerProvider(resource=resource)
    trace_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(trace_provider)

    # ==========================================================================
    # 2. Metrics
    # ==========================================================================
    metric_exporter = OTLPMetricExporter(endpoint=f"{otlp_endpoint}/v1/metrics")
    metric_reader = PeriodicExportingMetricReader(metric_exporter, export_interval_millis=60000)
    meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])
    metrics.set_meter_provider(meter_provider)

    # ==========================================================================
    # 3. Logs
    # ==========================================================================
    log_exporter = OTLPLogExporter(endpoint=f"{otlp_endpoint}/v1/logs")
    logger_provider = LoggerProvider(resource=resource)
    logger_provider.add_log_record_processor(BatchLogRecordProcessor(log_exporter))
    _logs.set_logger_provider(logger_provider)

    # Create handler for attaching to Python loggers
    _otel_log_handler = LoggingHandler(level=logging.DEBUG, logger_provider=logger_provider)

    # ==========================================================================
    # 4. Auto-instrumentation (non-Flask)
    # ==========================================================================
    # SQLAlchemy instrumentation
    SQLAlchemyInstrumentor().instrument()

    # Redis instrumentation
    RedisInstrumentor().instrument()

    # Celery instrumentation
    CeleryInstrumentor().instrument()

    # Logging instrumentation (adds trace_id, span_id to log records)
    LoggingInstrumentor().instrument(set_logging_format=True)

    _initialized = True


def get_otel_log_handler() -> LoggingHandler | None:
    """Get the OTel logging handler for attaching to loggers.

    Returns:
        The OTel LoggingHandler if initialized, None otherwise.
    """
    return _otel_log_handler


def instrument_flask_app(app) -> None:
    """Instrument a Flask app for tracing.

    This must be called after app creation, especially when using
    Gunicorn which forks workers after the global instrumentation
    is set up.

    Args:
        app: Flask application instance.
    """
    FlaskInstrumentor().instrument_app(app, excluded_urls="/api/health,/health")


def get_tracer(name: str) -> trace.Tracer:
    """Get a tracer for creating custom spans.

    Args:
        name: Name of the tracer (typically __name__).

    Returns:
        OpenTelemetry Tracer instance.
    """
    return trace.get_tracer(name)


def get_meter(name: str) -> metrics.Meter:
    """Get a meter for creating custom metrics.

    Args:
        name: Name of the meter (typically __name__).

    Returns:
        OpenTelemetry Meter instance.
    """
    return metrics.get_meter(name)
